File: unix/menu.c
// 23 april 2015
#include "uipriv_unix.h"

static GArray *menus = NULL;
static gboolean menusFinalized = FALSE;
static gboolean hasQuit = FALSE;
static gboolean hasPreferences = FALSE;
static gboolean hasAbout = FALSE;

struct uiMenu {
	char *name;
	GArray *items;					// []*uiMenuItem
	gboolean ischild;
};

struct uiMenuItem {
	char *name;
	int type;
	void (*onClicked)(uiMenuItem *, uiWindow *, void *);
	void *onClickedData;
	GType gtype;					// template for new instances; kept in sync with everything else
	gboolean disabled;
	gboolean checked;
	GArray *menuItems;			// []*menuItemWindow
	uiMenu *popupchild;
};

struct menuItemWindow {
	GtkMenuItem *item;
	uiWindow *w;
	gulong signal;
};

enum {
	typeRegular,
	typeCheckbox,
	typeQuit,
	typePreferences,
	typeAbout,
	typeSeparator,
	typeSubmenu,
};

// we do NOT want programmatic updates to raise an ::activated signal
static void singleSetChecked(GtkCheckMenuItem *menuitem, gboolean checked, gulong signal)
{
	g_signal_handler_block(menuitem, signal);
	gtk_check_menu_item_set_active(menuitem, checked);
	g_signal_handler_unblock(menuitem, signal);
}

static void setChecked(uiMenuItem *item, gboolean checked)
{
	struct menuItemWindow *w;
	guint i;

	item->checked = checked;
	for (i = 0; i < item->menuItems->len; i++) {
		w = g_array_index(item->menuItems, struct menuItemWindow *, i);
		singleSetChecked(GTK_CHECK_MENU_ITEM(w->item), item->checked, w->signal);
	}
}

static void onClicked(GtkMenuItem *menuitem, gpointer data)
{
	uiMenuItem *item = uiMenuItem(data);
	struct menuItemWindow *w;
	guint i;

	// we need to manually update the checked states of all menu items if one changes
	// notice that this is getting the checked state of the menu item that this signal is sent from
	if (item->type == typeCheckbox)
		setChecked(item, gtk_check_menu_item_get_active(GTK_CHECK_MENU_ITEM(menuitem)));

	for (i = 0; i < item->menuItems->len; i++) {
		w = g_array_index(item->menuItems, struct menuItemWindow *, i);
		if (w->item == menuitem)
			break;
	}

	(*(item->onClicked))(item, w->w, item->onClickedData);
}

static void defaultOnClicked(uiMenuItem *item, uiWindow *w, void *data)
{
	// do nothing
}

static void onQuitClicked(uiMenuItem *item, uiWindow *w, void *data)
{
	if (uiprivShouldQuit())
		uiQuit();
}

static void menuItemEnableDisable(uiMenuItem *item, gboolean enabled)
{
	struct menuItemWindow* w;
	guint i;

	item->disabled = !enabled;

	for (i = 0; i<item->menuItems->len; i++) {
		w = g_array_index(item->menuItems, struct menuItemWindow *, i);
		gtk_widget_set_sensitive(GTK_WIDGET(w->item), enabled);
	}
}

void uiMenuItemEnable(uiMenuItem *item)
{
	menuItemEnableDisable(item, TRUE);
}

void uiMenuItemDisable(uiMenuItem *item)
{
	menuItemEnableDisable(item, FALSE);
}

void uiMenuItemOnClicked(uiMenuItem *item, void (*f)(uiMenuItem *, uiWindow *, void *), void *data)
{
	if (item->type == typeQuit)
		uiprivUserBug("You cannot call uiMenuItemOnClicked() on a Quit item; use uiOnShouldQuit() instead.");
	if (item->type == typeSubmenu)
		uiprivUserBug("You cannot call uiMenuItemOnClicked() on an item with a subMenu.");
	item->onClicked = f;
	item->onClickedData = data;
}

int uiMenuItemChecked(uiMenuItem *item)
{
	return item->checked != FALSE;
}

void uiMenuItemSetChecked(uiMenuItem *item, int checked)
{
	gboolean c;

	// use explicit values
	c = FALSE;
	if (checked)
		c = TRUE;
	setChecked(item, c);
}

static uiMenuItem *newItem(uiMenu *m, int type, const char *name)
{
	uiMenuItem *item;

	if (menusFinalized)
		uiprivUserBug("You cannot create a new menu item after menus have been finalized.");

	item = uiprivNew(uiMenuItem);

	g_array_append_val(m->items, item);

	item->type = type;
	switch (item->type) {
	case typeQuit:
		item->name = g_strdup("Quit");
		break;
	case typePreferences:
		item->name = g_strdup("Preferences...");
		break;
	case typeAbout:
		item->name = g_strdup("About");
		break;
	case typeSeparator:
		break;
	default:
		item->name = g_strdup(name);
		break;
	}

	if (item->type == typeQuit) {
		// can't call uiMenuItemOnClicked() here
		item->onClicked = onQuitClicked;
		item->onClickedData = NULL;
	} else
		uiMenuItemOnClicked(item, defaultOnClicked, NULL);

	switch (item->type) {
	case typeCheckbox:
		item->gtype = GTK_TYPE_CHECK_MENU_ITEM;
		break;
	case typeSeparator:
		item->gtype = GTK_TYPE_SEPARATOR_MENU_ITEM;
		break;
	default:
		item->gtype = GTK_TYPE_MENU_ITEM;
		break;
	}

	item->menuItems = g_array_new(FALSE, TRUE, sizeof (struct menuItemWindow *));
	item->popupchild = NULL;

	return item;
}

uiMenuItem *uiMenuAppendSubmenu(uiMenu *m, uiMenu* child)
{
	uiMenuItem *item;

	if (menusFinalized)
		uiprivUserBug("You cannot create a new menu item after menus have been finalized.");

	item = uiprivNew(uiMenuItem);

	g_array_append_val(m->items, item);

	item->type = typeSubmenu;
	item->name = g_strdup(child->name);
	item->gtype = GTK_TYPE_MENU_ITEM;
	item->menuItems = g_array_new(FALSE, TRUE, sizeof (struct menuItemWindow *));
	item->popupchild = child;
	child->ischild = TRUE;

	return item;
}

uiMenuItem *uiMenuAppendItem(uiMenu *m, const char *name)
{
	return newItem(m, typeRegular, name);
}

uiMenuItem *uiMenuAppendCheckItem(uiMenu *m, const char *name)
{
	return newItem(m, typeCheckbox, name);
}

uiMenuItem *uiMenuAppendQuitItem(uiMenu *m)
{
	if (hasQuit)
		uiprivUserBug("You cannot have multiple Quit menu items in the same program.");
	hasQuit = TRUE;
	newItem(m, typeSeparator, NULL);
	return newItem(m, typeQuit, NULL);
}

uiMenuItem *uiMenuAppendPreferencesItem(uiMenu *m)
{
	if (hasPreferences)
		uiprivUserBug("You cannot have multiple Preferences menu items in the same program.");
	hasPreferences = TRUE;
	newItem(m, typeSeparator, NULL);
	return newItem(m, typePreferences, NULL);
}

uiMenuItem *uiMenuAppendAboutItem(uiMenu *m)
{
	if (hasAbout)
		uiprivUserBug("You cannot have multiple About menu items in the same program.");
	hasAbout = TRUE;
	newItem(m, typeSeparator, NULL);
	return newItem(m, typeAbout, NULL);
}

void uiMenuAppendSeparator(uiMenu *m)
{
	newItem(m, typeSeparator, NULL);
}

uiMenu *uiNewMenu(const char *name)
{
	uiMenu *m;

	if (menusFinalized)
		uiprivUserBug("You cannot create a new menu after menus have been finalized.");
	if (menus == NULL)
		menus = g_array_new(FALSE, TRUE, sizeof (uiMenu *));

	m = uiprivNew(uiMenu);

	g_array_append_val(menus, m);

	m->name = g_strdup(name);
	m->items = g_array_new(FALSE, TRUE, sizeof (uiMenuItem *));
	m->ischild = FALSE;

	return m;
}

static void appendMenuItem(GtkMenuShell *submenu, uiMenuItem *item, uiWindow *w)
{
	GtkWidget *menuitem;
	gulong signal;
	struct menuItemWindow *ww;

	uiMenu *child;
	guint j;
	GtkWidget *childSubmenu;

	menuitem = g_object_new(item->gtype, NULL);
	signal = 0;
	if (item->name != NULL)
		gtk_menu_item_set_label(GTK_MENU_ITEM(menuitem), item->name);
	if (item->type != typeSeparator && item->type != typeSubmenu) {
		signal = g_signal_connect(menuitem, "activate", G_CALLBACK(onClicked), item);
		gtk_widget_set_sensitive(menuitem, !item->disabled);
		if (item->type == typeCheckbox)
			singleSetChecked(GTK_CHECK_MENU_ITEM(menuitem), item->checked, signal);
	}
	gtk_menu_shell_append(submenu, menuitem);
	ww = uiprivNew(struct menuItemWindow);
	ww->item = GTK_MENU_ITEM(menuitem);
	ww->w = w;
	ww->signal = signal;
	g_array_append_val(item->menuItems, ww);

	if (item->popupchild != NULL) {
		child = item->popupchild;
		childSubmenu = gtk_menu_new();
		gtk_menu_item_set_submenu(GTK_MENU_ITEM(menuitem), childSubmenu);
		for (j = 0; j < child->items->len; j++)
			appendMenuItem(GTK_MENU_SHELL(childSubmenu), g_array_index(child->items, uiMenuItem *, j), w);
	}
}

GtkWidget *uiprivMakeMenubar(uiWindow *w)
{
	GtkWidget *menubar;
	guint i, j;
	uiMenu *m;
	GtkWidget *menuitem;
	GtkWidget *submenu;

	menusFinalized = TRUE;

	menubar = gtk_menu_bar_new();

	if (menus != NULL)
		for (i = 0; i < menus->len; i++) {
			m = g_array_index(menus, uiMenu *, i);
			if (m->ischild) continue;
			menuitem = gtk_menu_item_new_with_label(m->name);
			submenu = gtk_menu_new();
			gtk_menu_item_set_submenu(GTK_MENU_ITEM(menuitem), submenu);
			for (j = 0; j < m->items->len; j++)
				appendMenuItem(GTK_MENU_SHELL(submenu), g_array_index(m->items, uiMenuItem *, j), w);
			gtk_menu_shell_append(GTK_MENU_SHELL(menubar), menuitem);
		}

	gtk_widget_set_hexpand(menubar, TRUE);
	gtk_widget_set_halign(menubar, GTK_ALIGN_FILL);
	return menubar;
}

void uiprivFreeMenubar(uiWindow *w)
{
	guint i, j, windowIndex;
	uiMenu *m;
	uiMenuItem *item;
	struct menuItemWindow *ww;

	if (menus == NULL)
		return;

	// get the index of the uiWindow to free, assume that the first uiMenu has a uiMenuItem
	// TODO: is it safe to assume that? should we add a userBug if a uiMenu is empty?
	m = g_array_index(menus, uiMenu *, 0);
	item = g_array_index(m->items, uiMenuItem *, 0);
	for (i = 0; i < item->menuItems->len; i++)	{
		ww = g_array_index(item->menuItems, struct menuItemWindow *, i);
		if (ww->w == w) {
			windowIndex = i;
			break;
		}
	}

	for (i = 0; i < menus->len; i++) {
		m = g_array_index(menus, uiMenu *, i);
		for (j = 0; j < m->items->len; j++) {
			item = g_array_index(m->items, uiMenuItem *, j);
			ww = g_array_index(item->menuItems, struct menuItemWindow *, windowIndex);
			uiprivFree(ww);
			g_array_remove_index_fast(item->menuItems, windowIndex);
		}
	}
	// no need to worry about destroying any widgets; destruction of the window they're in will do it for us
}

void uiprivUninitMenus(void)
{
	uiMenu *m;
	uiMenuItem *item;
	guint i, j;

	if (menus == NULL)
		return;
	for (i = 0; i < menus->len; i++) {
		m = g_array_index(menus, uiMenu *, i);
		g_free(m->name);
		for (j = 0; j < m->items->len; j++) {
			item = g_array_index(m->items, uiMenuItem *, j);
			if (item->menuItems->len != 0)
				// TODO is this really a userbug()?
				uiprivImplBug("menu item %p (%s) still has uiWindows attached; did you forget to destroy some windows?", item, item->name);
			g_free(item->name);
			g_array_free(item->menuItems, TRUE);
			uiprivFree(item);
		}
		g_array_free(m->items, TRUE);
		uiprivFree(m);
	}
	g_array_free(menus, TRUE);
}

File: darwin/menu.m
// 28 april 2015
#import "uipriv_darwin.h"

struct uiMenu {
	NSMenu *menu;
	NSMenuItem *item;
};

struct uiMenuItem {
	uiprivMenuItem *item;
	int type;
	BOOL disabled;
	void (*onClicked)(uiMenuItem *, uiWindow *, void *);
	void *onClickedData;
};

enum uiprivMenuItemType {
	typeRegular,
	typeCheckbox,
	typeQuit,
	typePreferences,
	typeAbout,
};

@interface uiprivMenu : NSMenu {
@public
	uiMenu *menu;
}
@end

@implementation uiprivMenu
- (id)initWithTitle:(NSString *)title uiMenu:(uiMenu *)m
{
	self = [super initWithTitle:title];
	if (self) {
		self->menu = m;
	}
	return self;
}
@end

@implementation uiprivMenuItem
- (id)initWithTitle:(NSString *)title uiMenuItem:(uiMenuItem *)i
{
	self = [super initWithTitle:title action:@selector(onClicked:) keyEquivalent:@""];
	if (self) {
		self->item = i;

		[self setTarget:self];
	}
	return self;
}

- (IBAction)onClicked:(id)sender
{
	switch (self->item->type) {
	case typeQuit:
		if (uiprivShouldQuit())
			uiQuit();
		return;
	case typeCheckbox:
		uiMenuItemSetChecked(self->item, !uiMenuItemChecked(self->item));
		// fall through
	default:
		// System menu items that may have no user callback (Preferences/About)
		if (self->item == NULL)
			break;
		// use the key window as the source of the menu event; it's the active window
		(*(self->item->onClicked))(self->item, uiprivWindowFromNSWindow([uiprivNSApp() keyWindow]),
			self->item->onClickedData);
		break;
	}
}

- (void)setUiMenuItem:(uiMenuItem *)i
{
	self->item = i;
}

@end

@implementation uiprivMenuManager

- (id)init
{
	self = [super init];
	if (self) {
		self->hasQuit = NO;
		self->hasPreferences = NO;
		self->hasAbout = NO;
		self->finalized = NO;
	}
	return self;
}

- (BOOL)finalized
{
	return self->finalized;
}

- (void)finalize
{
	self->finalized = YES;
}

- (void)dealloc
{
	uiprivUninitMenus();
	[super dealloc];
}

- (void)register:(enum uiprivMenuItemType)type
{
	switch (type) {
	case typeQuit:
		if (self->hasQuit)
			uiprivUserBug("You can't have multiple Quit menu items in one program.");
		self->hasQuit = YES;
		break;
	case typePreferences:
		if (self->hasPreferences)
			uiprivUserBug("You can't have multiple Preferences menu items in one program.");
		self->hasPreferences = YES;
		break;
	case typeAbout:
		if (self->hasAbout)
			uiprivUserBug("You can't have multiple About menu items in one program.");
		self->hasAbout = YES;
		break;
	}
}

// on OS X there are two ways to handle menu items being enabled or disabled: automatically and manually
// unfortunately, the application menu requires automatic menu handling for the Hide, Hide Others, and Show All items to work correctly
// therefore, we have to handle enabling of the other options ourselves
- (BOOL)validateMenuItem:(NSMenuItem *)item
{
	// disable the special items if they aren't present
	if (item == self.quitItem && !self->hasQuit)
		return NO;
	if (item == self.preferencesItem && !self->hasPreferences)
		return NO;
	if (item == self.aboutItem && !self->hasAbout)
		return NO;
	// then poll the item's enabled/disabled state
	if ([item isKindOfClass:[uiprivMenuItem class]]) {
		uiprivMenuItem *mi = (uiprivMenuItem *)item;
		return !mi->item->disabled;
	}
	return NO;
}

// Cocoa constructs the default application menu by hand for each program; that's what MainMenu.[nx]ib does
- (void)buildApplicationMenu:(NSMenu *)menubar
{
	NSString *appName;
	NSMenuItem *appMenuItem;
	NSMenu *appMenu;
	NSMenuItem *item;
	uiprivMenuItem *pitem;
	NSString *title;
	NSMenu *servicesMenu;

	// note: no need to call setAppleMenu: on this anymore; see https://developer.apple.com/library/mac/releasenotes/AppKit/RN-AppKitOlderNotes/#X10_6Notes
	appName = [[NSProcessInfo processInfo] processName];
	appMenuItem = [[[NSMenuItem alloc] initWithTitle:appName action:NULL keyEquivalent:@""] autorelease];
	appMenu = [[[NSMenu alloc] initWithTitle:appName] autorelease];
	[appMenuItem setSubmenu:appMenu];
	[menubar addItem:appMenuItem];

	// first is About
	title = [@"About " stringByAppendingString:appName];
	pitem = [[[uiprivMenuItem alloc] initWithTitle:title uiMenuItem:NULL] autorelease];
	[appMenu addItem:pitem];
	self.aboutItem = pitem;

	[appMenu addItem:[NSMenuItem separatorItem]];

	// next is Preferences
	pitem = [[[uiprivMenuItem alloc] initWithTitle:@"Preferences\u2026" uiMenuItem:NULL] autorelease];
	[appMenu addItem:pitem];
	self.preferencesItem = pitem;

	[appMenu addItem:[NSMenuItem separatorItem]];

	// next is Services
	item = [[[NSMenuItem alloc] initWithTitle:@"Services" action:NULL keyEquivalent:@""] autorelease];
	servicesMenu = [[[NSMenu alloc] initWithTitle:@"Services"] autorelease];
	[item setSubmenu:servicesMenu];
	[uiprivNSApp() setServicesMenu:servicesMenu];
	[appMenu addItem:item];

	[appMenu addItem:[NSMenuItem separatorItem]];

	// next are the three hiding options
	title = [@"Hide " stringByAppendingString:appName];
	item = [[[NSMenuItem alloc] initWithTitle:title action:@selector(hide:) keyEquivalent:@"h"] autorelease];
	// the .xib file says they go to -1 ("First Responder", which sounds wrong...)
	// to do that, we simply leave the target as nil
	[appMenu addItem:item];
	item = [[[NSMenuItem alloc] initWithTitle:@"Hide Others" action:@selector(hideOtherApplications:) keyEquivalent:@"h"] autorelease];
	[item setKeyEquivalentModifierMask:(NSAlternateKeyMask | NSCommandKeyMask)];
	[appMenu addItem:item];
	item = [[[NSMenuItem alloc] initWithTitle:@"Show All" action:@selector(unhideAllApplications:) keyEquivalent:@""] autorelease];
	[appMenu addItem:item];

	[appMenu addItem:[NSMenuItem separatorItem]];

	// and finally Quit
	// DON'T use @selector(terminate:) as the action; we handle termination ourselves
	title = [@"Quit " stringByAppendingString:appName];
	pitem = [[[uiprivMenuItem alloc] initWithTitle:title uiMenuItem:NULL] autorelease];
	[appMenu addItem:pitem];
	self.quitItem = pitem;
}

- (NSMenu *)makeMenubar
{
	NSMenu *menubar;

	menubar = [[[NSMenu alloc] initWithTitle:@""] autorelease];
	[self buildApplicationMenu:menubar];
	return menubar;
}

@end

static void defaultOnClicked(uiMenuItem *item, uiWindow *w, void *data)
{
	// do nothing
}

void uiMenuItemEnable(uiMenuItem *item)
{
	item->disabled = NO;
	// we don't need to explicitly update the menus here; they'll be updated the next time they're opened (thanks mikeash in irc.freenode.net/#macdev)
}

void uiMenuItemDisable(uiMenuItem *item)
{
	item->disabled = YES;
}

void uiMenuItemOnClicked(uiMenuItem *item, void (*f)(uiMenuItem *, uiWindow *, void *), void *data)
{
	if (item->type == typeQuit)
		uiprivUserBug("You can't call uiMenuItemOnClicked() on a Quit item; use uiOnShouldQuit() instead.");
	item->onClicked = f;
	item->onClickedData = data;
}

int uiMenuItemChecked(uiMenuItem *item)
{
	return [item->item state] != NSOffState;
}

void uiMenuItemSetChecked(uiMenuItem *item, int checked)
{
	NSInteger state;

	state = NSOffState;
	if ([item->item state] == NSOffState)
		state = NSOnState;
	[item->item setState:state];
}

static uiMenuItem *newItem(uiMenu *m, int type, const char *name)
{
	@autoreleasepool {

	uiMenuItem *item;

	if ([uiprivAppDelegate().menuManager finalized])
		uiprivUserBug("You can't create a new menu item after menus have been finalized.");

	item = uiprivNew(uiMenuItem);

	item->type = type;
	switch (item->type) {
	case typeQuit:
		item->item = uiprivAppDelegate().menuManager.quitItem;
		[item->item setUiMenuItem:item];
		break;
	case typePreferences:
		item->item = uiprivAppDelegate().menuManager.preferencesItem;
		[item->item setUiMenuItem:item];
		break;
	case typeAbout:
		item->item = uiprivAppDelegate().menuManager.aboutItem;
		[item->item setUiMenuItem:item];
		break;
	default:
		item->item = [[uiprivMenuItem alloc] initWithTitle:uiprivToNSString(name) uiMenuItem:item];
		[m->menu addItem:item->item];
		break;
	}

	[uiprivAppDelegate().menuManager register:item->type];
	// typeQuit is handled via uiprivShouldQuit()
	if (item->type != typeQuit)
		uiMenuItemOnClicked(item, defaultOnClicked, NULL);

	return item;

	} // @autoreleasepool
}

uiMenuItem *uiMenuAppendItem(uiMenu *m, const char *name)
{
	return newItem(m, typeRegular, name);
}

uiMenuItem *uiMenuAppendCheckItem(uiMenu *m, const char *name)
{
	return newItem(m, typeCheckbox, name);
}

uiMenuItem *uiMenuAppendQuitItem(uiMenu *m)
{
	return newItem(m, typeQuit, NULL);
}

uiMenuItem *uiMenuAppendPreferencesItem(uiMenu *m)
{
	return newItem(m, typePreferences, NULL);
}

uiMenuItem *uiMenuAppendAboutItem(uiMenu *m)
{
	return newItem(m, typeAbout, NULL);
}

void uiMenuAppendSeparator(uiMenu *m)
{
	[m->menu addItem:[NSMenuItem separatorItem]];
}

uiMenu *uiNewMenu(const char *name)
{
	@autoreleasepool {

	uiMenu *m;

	if ([uiprivAppDelegate().menuManager finalized])
		uiprivUserBug("You can't create a new menu after menus have been finalized.");

	m = uiprivNew(uiMenu);

	m->menu = [[uiprivMenu alloc] initWithTitle:uiprivToNSString(name) uiMenu:m];
	// use automatic menu item enabling for all menus for consistency's sake

	m->item = [[NSMenuItem alloc] initWithTitle:uiprivToNSString(name) action:NULL keyEquivalent:@""];
	[m->item setSubmenu:m->menu];

	[[uiprivNSApp() mainMenu] addItem:m->item];

	return m;

	} // @autoreleasepool
}

void uiprivFinalizeMenus(void)
{
	[uiprivAppDelegate().menuManager finalize];
}

void uiprivUninitMenus(void)
{
	NSMenuItem *mi;
	NSMenu *sm;
	NSMenuItem *smi;

	for (mi in [[uiprivNSApp() mainMenu] itemArray]) {
		if ([mi hasSubmenu]) {
			sm = [mi submenu];
			for (smi in [sm itemArray]) {
				if ([smi isKindOfClass:[uiprivMenuItem class]]) {
					uiprivMenuItem *x = (uiprivMenuItem *)smi;
					if (x->item != NULL)
						uiprivFree(x->item);
				}
			}
			if ([sm isKindOfClass:[uiprivMenu class]]) {
				uiprivMenu *x = (uiprivMenu *)sm;
				uiprivFree(x->menu);
			}
		}
	}
}
